control/HTTPCacheControl.php

```diff
@@ -0,0 +1,183 @@
+<?php
+
+/**
+ * Class HTTPCacheControl
+ *
+ *
+ */
+class HTTPCacheControl extends SS_Object {
+
+	/**
+	 * @var static
+	 */
+	private static $inst;
+
+	private $state = array();
+
+	/**
+	 * A list of allowed cache directives for HTTPResponses
+	 *
+	 * This doesn't include any experimental directives,
+	 * use the config system to add to these if you want to enable them
+	 *
+	 * @config
+	 * @var array
+	 */
+	private static $allowed_directives = array(
+		'public',
+		'private',
+		'no-cache',
+		'max-age',
+		's-maxage',
+		'must-revalidate',
+		'proxy-revalidate',
+		'no-store',
+		'no-transform',
+	);
+
+	public function setDirective($directive, $value = null)
+	{
+		$allowedDirectives = $this->config()->get('allowed_directives');
+		$directive = strtolower($directive);
+		if (in_array($directive, $allowedDirectives)) {
+			$this->state[$directive] = $value;
+		} else {
+			throw new InvalidArgumentException('Directive ' . $directive . ' is not allowed');
+		}
+		return $this;
+	}
+
+	public function setDirectivesFromArray($directives)
+	{
+		foreach ($directives as $directive => $value) {
+			if (is_null($value)) {
+				$this->removeDirective($directive);
+			} else {
+				if ($value && (is_bool($value) || strtolower($value) === 'true')) {
+					$value = null;
+				}
+				$this->setDirective($directive, $value);
+			}
+		}
+		return $this;
+	}
+
+	public function removeDirective($directive)
+	{
+		unset($this->state[strtolower($directive)]);
+		return $this;
+	}
+
+	public function hasDirective($directive)
+	{
+		return array_key_exists(strtolower($directive), $this->state);
+	}
+
+	public function getDirective($directive)
+	{
+		if ($this->hasDirective($directive)) {
+			return $this->state[strtolower($directive)];
+		}
+		return false;
+	}
+
+	public function setNoStore($noStore = true)
+	{
+		if ($noStore) {
+			$this->setDirective('no-store');
+		} else {
+			$this->removeDirective('no-store');
+		}
+		return $this;
+	}
+
+	public function setNoCache($noCache = true)
+	{
+		if ($noCache) {
+			$this->setDirective('no-cache');
+		} else {
+			$this->removeDirective('no-cache');
+		}
+		return $this;
+	}
+
+	public function setPublic()
+	{
+		$this->setDirective('public');
+		$this->removeDirective('private');
+		return $this;
+	}
+
+	public function setPrivate()
+	{
+		$this->setDirective('private');
```

Comment:
This should also unset `max-age` and `s-maxage`, right? I doubt it would overrule `cache-control: private` in HTTP caches, but seems confusing to send those mixed signals. Is there any point to the `vary` header when using `cache-control: private`?

Reply:
No, `max-age` and `s-maxage` are still valid directives for a private cache.

All the `private` directive does is tell intermidiary proxies/cache stores they should not cache this response as it contains private data just for the end user; so browsers can cache it.